lib/anonymousNaming.js

```diff
@@ -43,42 +49,170 @@ function assignUidToParameterSchemas(doc) {
+
+          // This example is working because `Car` and `Airplane` are root level schemas and gots its own ids:
+          //   properties:
+          //     vehicle:
+          //       type: array
+          //       x-parser-schema-id: mySchemaVehicleArray
+          //       items:
+          //         - $ref: "#/components/schemas/Car"
+          //         - $ref: "#/components/schemas/Airplane"
+
+          // but if the items are anonymous
+          //   properties:
+          //     vehicle:
+          //       type: array
+          //       x-parser-schema-id: mySchemaVehicleArray <-- collision
+          //       items:
+          //         - type: object
+          //           x-parser-schema-id: mySchemaVehicle
+          //           properties:
+          //             maxSpeed: 
+          //               type: string
+          //         - type: object
+          //           x-parser-schema-id: mySchemaVehicle <-- collision
+          //           properties:
+          //             maxAltitude: 
+          //               type: string
+          //
+          // would not work so we can not give code generator firendly id when `items`is an array.
+          return null;
+        }
+
+        return options.parentId + upperFirst(propertyName);
+      },
+      callback: (subSchema, propOrIndex, callbackType, options) => {
+        if (callbackType === SchemaIteratorCallbackType.END_SCHEMA) {
+          return;
+        }
+
+        if (propOrIndex === null &&
+          options.parentType === 'array' // Special edge case, here we need to inheritance the id to the childs.
+        ) {
+          propOrIndex = options.parentId;
+        }
+
+        if (propOrIndex === null) {
+          return false;
+        }
+
+        if (!(subSchema.type() === 'object' ||
+          subSchema.type() === 'array' ||
+          (subSchema.type() === 'string' && subSchema.enum()) ||
+          !subSchema.type() // oneOf, anyOf, allOf dont have a type.
+        )) {
+          // Simple types dont need generated ids.
+          return false;
+        }
+
+        options.parentId = propOrIndex;
+        options.parentType = subSchema.type();
+        options.level++;
+
+        if (!subSchema.$id()) {
+          if (subSchema.json()[String(xParserSchemaIdLevel)] && subSchema.json()[String(xParserSchemaIdLevel)] < options.level) {
+            // After deref we get a tree, always prefer names of parent root schema definitions.
+
+            // The usage of $ref will lead to having the same schema multiple times in the tree caused by deref.
+            // By walking through the tree, each level deper in the tree gets a higher xParserSchemaIdLevel
+            // So if we find a genereated id with a higher xParserSchemaIdLevel the the current, this id was generated because the current schema was used via $ref as subelement in another schema.
+            // and we prefer the generated id that is clother to a rood node in tree.
```

Comment:
```suggestion
            // and we prefer the generated id that is closer to a root node in the tree.
```

Is that what you meant?